driver/src/orderbook/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,317 @@
+            .user_order_counts
+            .get(&last_order_user)
+            .expect("user has order but no order count");
+        number_of_added_orders
+    }
+
+    fn apply_auction_data(&mut self, packed_auction_bytes: &[u8]) {
+        let auction_elements = self.parse_auction_elements(packed_auction_bytes);
+        self.apply_auction_elements_to_account_state(auction_elements.iter());
+        self.orders.extend(
+            auction_elements
+                .into_iter()
+                .map(|auction_element| auction_element.order),
+        );
+    }
+
+    fn apply_auction_elements_to_account_state<'a, Iter>(&mut self, auction_elements: Iter)
+    where
+        Iter: Iterator<Item = &'a StableXAuctionElement>,
+    {
+        for element in auction_elements.into_iter() {
+            self.account_state.modify_balance(
+                element.order.account_id,
+                element.order.sell_token,
+                |x| {
+                    if *x == 0 {
+                        *x = element.sell_token_balance
+                    } else {
+                        assert_eq!(
+                            *x,
+                            element.sell_token_balance,
+                            "got order which sets user {}'s sell token {} \
+                            balance to {} but sell_token_balance has already \
+                            been set to {}",
+                            element.order.account_id,
+                            element.order.sell_token,
+                            element.sell_token_balance,
+                            *x
+                        );
+                    }
+                },
+            );
+        }
+    }
+
+    fn parse_auction_elements(
+        &mut self,
+        packed_auction_bytes: &[u8],
+    ) -> Vec<StableXAuctionElement> {
+        assert_eq!(
+            packed_auction_bytes.len() % AUCTION_ELEMENT_WIDTH,
+            0,
+            "Each auction should be packed in {} bytes",
+            AUCTION_ELEMENT_WIDTH
+        );
+
+        // Workaround for borrow checker that would complain that the map
+        // closure borrows self as mutable while at the same time the filter
+        // closure borrows self as immutable because of using `self.index`.
+        let index = self.index;
+        packed_auction_bytes
+            .chunks(AUCTION_ELEMENT_WIDTH)
+            .map(|chunk| {
+                let mut chunk_array = [0; AUCTION_ELEMENT_WIDTH];
+                chunk_array.copy_from_slice(chunk);
+                let mut result = StableXAuctionElement::from_bytes(&chunk_array);
+                let order_counter = self
+                    .user_order_counts
+                    .entry(result.order.account_id)
+                    .or_insert(0);
+                result.order.id = *order_counter as u16;
+                *order_counter += 1;
+                result
+            })
+            .filter(|x| x.in_auction(index) && x.order.sell_amount > 0)
```

Comment:
I think technically, since the contract can only have a minimum executed amounts as 10^4, this inequality could be the same.

Reply:
Here is the point where you filter this reducing the amount of orders taken from the page.